### PR TITLE
ci: auto label PR

### LINKED ISSUE #2945: add Labels to github PR
It's a little bit nosying to watch all PR.

Add labels to pr would help to filter PR like "translation: CN"

### PR DESCRIPTION
**Pull request by Trim21**
close #2945

### CODE REVIEW

File: .github/scripts/auto-label.cjs [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+  '.github/scripts/': 'ci',
+  'ui/config/': 'config',
+  'ui/dps/': 'dps',
+  'ui/eureka/': 'eureka',
+  'ui/fisher/': 'fishing🎣',
+  'ui/jobs/': 'jobs',
+  'ui/oopsyraidsy/': 'oopsy',
+  'ui/pullcounter/': 'pullcounter',
+  'ui/radar/': 'radar',
+  'ui/raidboss/': 'raidboss',
```
Comment (quisquous) [inline .github/scripts/auto-label.cjs:34]: Were you still planning to add a raidemulator label as we discussed on the other PR, or should that be a followup PR?

Comment (Trim21) [reply .github/scripts/auto-label.cjs:34]: I'm waiting for a conclusion or a decision. I think we didn't get one in previous PR.

Comment (quisquous) [reply .github/scripts/auto-label.cjs:34]: I think there weren't strong opinions.  I think separating sounds good.  In other words, if something is only a raidemulator change then it gets a raidemulator label and not a raidboss label.

Comment (Trim21) [reply .github/scripts/auto-label.cjs]: just done

**Comment by Trim21** [comment]
There is a problem: label `test` and `ui/test/` are both `test`

**Comment by quisquous** [comment]
> There is a problem: `test/` and `ui/test/` are both labeled `test`
> 
> maybe we still need a `module` prefix

I think this is fine.  `ui/test/` is a manual test / debug info imo.  I would also be ok with not giving `ui/test/` a label at all if you thought that was better.  I think either of those are better than really long labels.

**Comment by Trim21** [comment]
ok I'll keep it this way.

**Comment by Trim21** [comment]
I think this PR is ready to be merged,  is there any more questions?

**Comment by quisquous** [comment]
### better printing

Can I suggest more consistent output so it'll be easier to figure out in the future from a log how labels got applied?

```diff
giff --git a/.github/scripts/auto-label.cjs b/.github/scripts/auto-label.cjs
index 02737a69..6db48fc3 100644
--- a/.github/scripts/auto-label.cjs
+++ b/.github/scripts/auto-label.cjs
@@ -94,7 +94,6 @@ const getLabels = async (github, owner, repo, pullNumber) => {
   }).map((f) => f()));
 
   const changedLang = getTimelineReplaceChanges(changedFilesContent);
-  console.log(`changed timelineReplace ${changedLang}`);
   changedLang.push(...nonNullUnique(lodash.flatten(changedFiles.map((f) => {
     if (['.js', '.ts'].includes(path.extname(f.filename)))
       return parseChangedLang(f.patch);
@@ -103,8 +102,10 @@ const getLabels = async (github, owner, repo, pullNumber) => {
   // by file path
   const changedModule = nonNullUnique(changedFiles.map((f) => {
     for (const [prefix, label] of Object.entries(prefixLabelMap)) {
-      if (f.filename.startsWith(prefix))
+      if (f.filename.startsWith(prefix)) {
+        console.log(`label: ${label} [filename] (${f.filename})`);
         return label;
+      }
     }
   }));
 
@@ -140,8 +141,12 @@ const parseChangedLang = (patch) => {
   for (const lang of ['cn', 'de', 'ja', 'fr']) {
     const pattern = new RegExp(`^\s*[+].*(?:${lang}|'${lang}'): `);
     for (const line of patch.split('\n')) {
-      if (pattern.test(line))
+      if (pattern.test(line)) {
+        // TODO: it'd be nice to add the file/line number here.
+        if (!set.has(lang))
+          console.log(`label: ${lang} [translation]`);
         set.add(lang);
+      }
     }
   }
   return Array.from(set.values());
@@ -181,11 +186,13 @@ const getTimelineReplaceChanges = (changedFiles) => {
       return;
 
     if (path.extname(file.filename) === '.js') {
-      const from = getTimelineReplace(file.from);
-      const to = getTimelineReplace(file.to);
-      for (const lang of lodash.uniq([...Object.keys(from || {}), ...Object.keys(to || {})])) {
-        if (!lodash.isEqual(from[lang], to[lang]))
+      const from = getTimelineReplace(file.from) || {};
+      const to = getTimelineReplace(file.to) || {};
+      for (const lang of lodash.uniq([...Object.keys(from), ...Object.keys(to)])) {
+        if (!lodash.isEqual(from[lang], to[lang])) {
+          console.log(`label: ${lang} [timelineReplace] (${file.filename})`);
           s.add(lang);
+        }
       }
     }
   });
@@ -228,7 +235,7 @@ const run = async () => {
 
   const labels = await getLabels(octokit, owner, repo, pullNumber);
   if (labels && labels.length) {
-    console.log(`set ${labels}`);
+    console.log(`apply: ${JSON.stringify(labels)}`);
     const res = await octokit.request(`PUT /repos/${owner}/${repo}/issues/${pullNumber}/labels`, { labels });
     if (res.status !== 200)
       console.log(res);
```

^ This also includes a fix in timelineReplace for PR=2966 where `from` or `to` could be undefined.

This prints out things like this:

```
PR_NUMBER=2966 node .github/scripts/auto-label.cjs
label: de [timelineReplace] (ui/raidboss/data/03-hw/raid/a10s.js)
label: de [timelineReplace] (ui/raidboss/data/04-sb/alliance/ridorana_lighthouse.js)
label: de [timelineReplace] (ui/raidboss/data/04-sb/raid/o2s.js)
label: de [timelineReplace] (ui/raidboss/data/04-sb/trial/lakshmi-ex.js)
label: de [timelineReplace] (ui/raidboss/data/05-shb/trial/diamond_weapon-ex.js)
label: de [timelineReplace] (ui/raidboss/data/05-shb/trial/diamond_weapon.js)
label: de [timelineReplace] (ui/raidboss/data/05-shb/trial/emerald_weapon.js)
label: fr [timelineReplace] (ui/raidboss/data/05-shb/trial/emerald_weapon.js)
label: ja [timelineReplace] (ui/raidboss/data/05-shb/trial/emerald_weapon.js)
label: cn [timelineReplace] (ui/raidboss/data/05-shb/trial/emerald_weapon.js)
label: ko [timelineReplace] (ui/raidboss/data/05-shb/trial/emerald_weapon.js)
label: de [timelineReplace] (ui/raidboss/data/05-shb/trial/innocence-ex.js)
label: de [timelineReplace] (ui/raidboss/data/05-shb/ultimate/the_epic_of_alexander.js)
label: de [translation]
label: raidboss [filename] (ui/raidboss/data/03-hw/raid/a10s.js)
label: raidboss [filename] (ui/raidboss/data/04-sb/alliance/ridorana_lighthouse.js)
label: raidboss [filename] (ui/raidboss/data/04-sb/raid/o2s.js)
label: raidboss [filename] (ui/raidboss/data/04-sb/trial/lakshmi-ex.js)
label: raidboss [filename] (ui/raidboss/data/05-shb/eureka/zadnor.js)
label: raidboss [filename] (ui/raidboss/data/05-shb/trial/diamond_weapon-ex.js)
label: raidboss [filename] (ui/raidboss/data/05-shb/trial/diamond_weapon.js)
label: raidboss [filename] (ui/raidboss/data/05-shb/trial/emerald_weapon.js)
label: raidboss [filename] (ui/raidboss/data/05-shb/trial/innocence-ex.js)
label: raidboss [filename] (ui/raidboss/data/05-shb/ultimate/the_epic_of_alexander.js)
apply: ["raidboss","�DE","�FR","�JA","�CN","�KO","�DE"]
```

### errors

I wanted to look at what got listed for new PRs for raidboss files (that might include extra translations, but aren't really *translations* per se), but all of the new raidboss file PRs I tried had errors, e.g.:

```
PR_NUMBER=2944 node .github/scripts/auto-label.cjs
SyntaxError: Unexpected token (5:0)
    at _temp._raise (C:\Users\enne\cactbot\node_modules\@babel\parser\lib\index.js:748:17)
    at _temp.raiseWithData (C:\Users\enne\cactbot\node_modules\@babel\parser\lib\index.js:741:17)
    at _temp.raise (C:\Users\enne\cactbot\node_modules\@babel\parser\lib\index.js:735:17)
    at _temp.unexpected (C:\Users\enne\cactbot\node_modules\@babel\parser\lib\index.js:9097:16)
    at _temp.parseExprAtom (C:\Users\enne\cactbot\node_modules\@babel\parser\lib\index.js:10548:20)
    at _temp.parseExprAtom (C:\Users\enne\cactbot\node_modules\@babel\parser\lib\index.js:4759:20)
    at _temp.parseExprSubscripts (C:\Users\enne\cactbot\node_modules\@babel\parser\lib\index.js:10122:23)
    at _temp.parseUpdate (C:\Users\enne\cactbot\node_modules\@babel\parser\lib\index.js:10102:21)
    at _temp.parseMaybeUnary (C:\Users\enne\cactbot\node_modules\@babel\parser\lib\index.js:10091:17)
    at _temp.parseExprOps (C:\Users\enne\cactbot\node_modules\@babel\parser\lib\index.js:9961:23) {
  loc: Position { line: 5, column: 0 },
  pos: 8
```

^ same with 1677, 2734, 2846

**Comment by Trim21** [comment]
Could you give a patch directly? so I can apply it with `git apply`

**Review by quisquous — COMMENTED** [review]
> Could you give a patch directly? so I can apply it with `git apply`

[0001-add-better-printing.zip](https://github.com/quisquous/cactbot/files/6571974/0001-add-better-printing.zip)

Sure thing.  (I think you can just `patch -p1` the original diff too?)

**Comment by Trim21** [comment]
> > Could you give a patch directly? so I can apply it with `git apply`
> 
> [0001-add-better-printing.zip](https://github.com/quisquous/cactbot/files/6571974/0001-add-better-printing.zip)
> 
> Sure thing. (I think you can just `patch -p1` the original diff too?)

I don't know why I can't. I got `error: corrupt patch at line 7`

**Comment by Trim21** [comment]
`// TODO: it'd be nice to add the file/line number here.`

I'd rather not parse the patch file.

**Comment by quisquous** [comment]
> `// TODO: it'd be nice to add the file/line number here.`
> 
> I'd rather not to parse the patch file.

That's fair! I was leaving a TODO as a "this could be nice in the future if somebody wanted to" and definitely not a suggestion to do here.  I think this is close to landing if the "unexpected token" error can be fixed.

**Comment by Trim21** [comment]
I find out the reason, the base files is missing. I'll fix it now.

**Comment by Trim21** [comment]
done

**Comment by quisquous** [comment]
:tada:

**Comment by quisquous** [comment]
Also, thank you *soooo* much for taking the time to write this automated script.  I think this is going to be really useful and I don't really know enough to write any of this on my own.

**Comment by Trim21** [comment]
I just find that it failed on https://github.com/quisquous/cactbot/pull/2973...

**Comment by quisquous** [comment]
Do I need to fix the token permissions? What permissions did you add when you were testing?

**Comment by Trim21** [comment]
No, I need to update  the trigger event to `pull_request_target`

**Pull request merged**